### PR TITLE
fix #17740: `yii\helpers\BaseInflector::slug()` doesn't replace multip…

### LINKED ISSUE #17740: yii\helpers\BaseInflector::slug() doesn't shrink multiple replacement occurences to single
### What steps will reproduce the problem?
Inflector::slug(' _ _ remove excess _ _ replacements_', '_');
### What is the expected result?
remove_excess_replacements
### What do you get instead?
remove_excess_____replacements

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.16-2.0.30 since #15204
| PHP version      | 7.0.33
| Operating system | Debian 7

### PR DESCRIPTION
**Pull request by batyrmastyr**
…le replacement string occurrences to single one

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #17740

### CODE REVIEW

File: framework/helpers/BaseInflector.php [hunk 1]
```diff
@@ -485,6 +485,7 @@ public static function slug($string, $replacement = '-', $lowercase = true)
         }, $parts);
 
         $string = trim(implode($replacement, $replaced), $replacement);
+        $string = preg_replace('#'.preg_quote($replacement).'+#', $replacement, $string);
```
Comment (kamarton) [inline framework/helpers/BaseInflector.php]: > ... s a backwards compatibility break for Yii 2.

add an extra parameter to the method (only for yii2). @samdark ?

```php
public static function slug($string, $replacement = '-', $lowercase = true, $clean = false)
```

I would not implement the extra parameter in yii3.

Comment (samdark) [reply framework/helpers/BaseInflector.php]: I'd use extra preg_replace or external wrapper or your own implementation of Inflector instead. That's not something that could be considered Yii 2 bug.

Comment (batyrmastyr) [reply framework/helpers/BaseInflector.php]: It was introduced in Yii 2.0.16 and it doesn't look like intended change, so why it's not a bug? In terms of age - 2.0.0 - 2.0.16 is 4 years, 2.0.16 - now is 1.5 year.

Comment (samdark) [reply framework/helpers/BaseInflector.php]: Hmm... wait, it was doing that before 2.0.16?

Comment (samdark) [reply framework/helpers/BaseInflector.php]: Alright. That's bug.

**Comment by samdark** [comment]
@batyrmastyr would you please add a note about output change for UPGRADE document? Thanks.

**Comment by batyrmastyr** [comment]
@samdark should there be `clean` option with true or false or no option at all?
```Upgrade from Yii 2.0.30:```

- In case `$clean` defaults to `true`.

`yii\helpers\BaseInflector::slug()` now ensures that there is no repeating $replacement occurrences in slug, ex. 'slug-without-repeats' vs 'slug-with---repeats'. In case you relied on behavior, replace `Inflector` with your own implementation.
```
public static function slug($string, $replacement = '-', $lowercase = true, $clean = true)
{
   return parent::slug($string, $replacement, $lowercase, $clean)
}
Yii::$classMap['yii\helpers\Inflector'] = '@your/Inflector.php';
```
- In case `$clean` defaults to `false`. 

If you upgraded from Yii 2.0.16 and expect `yii\helpers\BaseInflector::slug()` to remove repeating $replacement occurrences, ex. 'slug-without-repeats' vs 'slug-with--repeats', replace `Inflector` with your own implementation.
```
public static function slug($string, $replacement = '-', $lowercase = true, $clean = true)
{
   return parent::slug($string, $replacement, $lowercase, $clean)
}
Yii::$classMap['yii\helpers\Inflector'] = '@your/Inflector.php';
```
- In case there is not option at all, there is nothing to write except apologies.

**Comment by samdark** [comment]
No option at all, I think since it's a regression.

**Comment by samdark** [comment]
> In case there is not option at all, there is nothing to write except apologies.

Kind of :) The info should be there about the fact of the change.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**